src/library/scala/runtime/BoxesRunTime.java

```diff
@@ -117,36 +117,29 @@ public static double unboxToDouble(Object d) {
     /* COMPARISON ... COMPARISON ... COMPARISON ... COMPARISON ... COMPARISON ... COMPARISON */
 
     public static boolean equals(Object x, Object y) {
-        if (x == y) return true;
-        return equals2(x, y);
+        return (x == y) || equalsNotSame(x, y);
     }
 
-    /** Since all applicable logic has to be present in the equals method of a ScalaNumber
-     *  in any case, we dispatch to it as soon as we spot one on either side.
-     */
-    public static boolean equals2(Object x, Object y) {
-        if (x instanceof java.lang.Number)
-            return equalsNumObject((java.lang.Number)x, y);
-        if (x instanceof java.lang.Character)
-            return equalsCharObject((java.lang.Character)x, y);
-        if (x == null)
-            return y == null;
-
-        return x.equals(y);
+    private static boolean equalsNotSame(Object x, Object y) {
+        return x != null && y != null && equalsNotSameOrNull(x, y);
```

Comment:
if they're not the same (`==`), but one of them is `null`, they're certainly not `equal`
(`scala.runtime.BoxesRunTime::equalsNotSame (22 bytes)   inline (hot)`)